data/crac/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/SimpleCrac.java

```diff
@@ -283,78 +217,49 @@ public void removeState(String id) {
+            return getState(contingency, instant);
         } else {
-            if (getInstant(instantId) != null) {
-                state = new SimpleState(Optional.empty(), getInstant(instantId));
-            } else {
-                throw new FaraoException(format(ADD_ELEMENT_TO_CRAC_ERROR_MESSAGE, instantId));
+            if (!contingencies.containsKey(contingency.getId())) {
+                throw new FaraoException(format(ADD_ELEMENT_TO_CRAC_ERROR_MESSAGE, contingency.getId()));
             }
+            State state = new PostContingencyState(getContingency(contingency.getId()), instant);
+            states.put(state.getId(), state);
+            return state;
         }
-        states.put(state.getId(), state);
-        return state;
     }
 
-    /**
-     * Add a state in the Crac object. When adding a state which is made of a Contingency and an Instant,
-     * these Contingency and Instant objects have to be independently present in the Crac as well. So if they
-     * are not, they have to be added as well.
-     * Then the State has to point on the good objects, meaning those which are present independently in the Crac.
-     * So in the end a new State object will be created with references on Contingency and Instant objects
-     * that have been already added to the Crac.
-     *
-     * @param state: state object that can be created from already existing Contingency and Instant object of the Crac
-     *             or not.
-     */
-    @Override
-    public void addState(State state) {
-        // If the two instants are strictly equals no need to add it
-        if (instants.values().stream().noneMatch(instant ->
-            instant.getId().equals(state.getInstant().getId()) && instant.getSeconds() == state.getInstant().getSeconds())
-        ) {
-            // Can thow FaraoException if this instant and already present instants are incompatible
-            addInstant(state.getInstant());
-        }
-        Instant instant = getInstant(state.getInstant().getId());
-
-        Optional<Contingency> stateContingency = state.getContingency();
-        Optional<Contingency> contingency;
-        if (stateContingency.isPresent()) {
-            if (contingencies.values().stream().noneMatch(stateContingency.get()::equals)) {
-                addContingency(stateContingency.get());
+    private State addState(State state) {
+        if (getState(state.getId()) == null) {
+            Optional<Contingency> optContingency = state.getContingency();
+            if (optContingency.isPresent()) {
+                Contingency contingency = optContingency.get();
+                if (getContingency(contingency.getId()) == null) {
+                    throw new FaraoException(format(ADD_ELEMENT_TO_CRAC_ERROR_MESSAGE, contingency.getId()));
+                } else {
+                    return addState(contingency, state.getInstant());
+                }
+            } else {
+                return addPreventiveState();
```

Comment:
Should we check that state.getId() is the same as our PreventiveState.getId()?
A user could call addState(state) with a state that has no contingency, and then expect getPreventiveState() and his state to be equal (ie have the same id)

Reply:
I tried something to avoid this kind of treatment. At least we assume explicitly that we can only handle PreventiveState -- that would always have the same ID -- and PostContingencyState -- that would always have a contingency defined.